CC-Widgets Packages setup

Shreyas281299 opened the pull request:
## COMPLETED https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-569253

## Changes
Adds basic setup for widget
### Functionalities
- Adds ability to build providers - yarn build:providers
- Adds ability to build widgets - yarn build:widgets
- Adds ability to build both at once (providers should be built first) yarn run build
- Adds ability to build samples, both react based samples and web-component- yarn run build:samples
- Adds ability to run react samples- yarn samples:serve-react
- Adds ability to run web-component samples - yarn samples:serve-wc
- Adds ability to run tests(tests are only setup in widget-cc-station-login)

### Testing
- Built providers
- Build widgets
- Build samples
- Ran samples
- Ran test for widget-cc-station-login

Shreyas281299 commented on .babelrc.js:
```diff
@@ -0,0 +1,7 @@
+module.exports = {
```
Created common babelrc which will be imported in each workspace. babelrc is required in each workspace to run jest tests

Shreyas281299 commented on .gitignore:
```diff
@@ -43,6 +43,7 @@ reports/
 !.env.default
 !package.json
 !package-lock.json
+!tsconfig.json
```
tsconfig was getting ignore, was causing issues.

Shreyas281299 commented on docs/react-samples/package.json:
```diff
@@ -0,0 +1,32 @@
+{
```
The sample app is also a workspace. This will allow us to easily link updated widgets locally.

Shreyas281299 replied:
All the files in this are demo, these will be updated when we add the actual implementation of these widgets. 
We can still run and see the basic providers and widgets working and sharing states.

Shreyas281299 commented on docs/react-samples/src/App.tsx:
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
```
Dummy code for a simple demo.

Shreyas281299 commented on docs/react-samples/webpack.config.js:
```diff
@@ -0,0 +1,38 @@
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+const {merge} = require('webpack-merge');
+
+const baseConfig = require('../../webpack.config');
```
Using base webpack config here also

Shreyas281299 commented on docs/web-component-samples/index.html:
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
```
This is a sample using web-components and NO react. Again dummy code to just show that widgets can be used as web-components

Shreyas281299 commented on packages/@webex/widget-cc-station-login/.babelrc.js:
```diff
@@ -0,0 +1,3 @@
+const baseConfig = require('../../../.babelrc');
```
using base babelrc config.

arun3528 reviewed (changes requested):
I dont think you have to follow the structure of packages/@webex structure here 

yarn workspace can just relay on the folder structure like widgets/cc-station-login etc 

the npm package where you are going to publish can be @webex/cc-station-login

The naming looks little messed up i know we did not follow this on the sdk side

arun3528 commented on .gitignore:
```diff
@@ -43,6 +43,7 @@ reports/
 !package.json
 !package-lock.json
+!tsconfig.json
 node_modules/
```
this should be /node_moidules/ once you add yarn i guess

Shreyas281299 replied:
node_modules/ is what is ignorig the node_modules inside each workspace. If I change it to /node_modules/ then its not ignoring node_modules.

arun3528 commented on packages/@webex/widget-cc-station-login/src/hooks.ts:
```diff
@@ -0,0 +1,3 @@
+export const useStationLogin = () => {
+  return {name: 'StationLogin'};
```
can we confirm the naming of this file as hooks if its going to be just a helper file for components should be name it utils or something as it gets confused with react hooks

if we going to import react stuff here i assume its ok

Shreyas281299 replied:
Changed to helper.ts

arun3528 commented on packages/@webex/widget-cc-station-login/src/station-login/index.tsx:
```diff
@@ -0,0 +1,25 @@
+const StationLogin: React.FunctionComponent = observer(() => {
+  const {loginState, setLoginState, ccSdk, isAvailable} = store;
+  const result = useStationLogin();
+
```
we will need to add some documentation here

mkesavan13 replied:
What kind of documentation are you looking for Arun? Just code comments or something like API reference?

arun3528 replied:
code comments in general

mkesavan13 replied:
This is just dummy code. We will add comments and documentation when we write the actual code

arun3528 commented on packages/@webex/widget-provider/package.json:
```diff
@@ -0,0 +1,42 @@
+{
```
should we rename this to something else , provider is too generic , should it be called widgets store or something ?

mkesavan13 replied:
`@webex/widgets-store` sounds good to me

Shreyas281299 replied:
Changed to widgets-store

mkesavan13 replied:
This is now changed to `@webex/cc-store` as a final call

arun3528 commented on packages/@webex/widget-provider/src/store.ts:
```diff
@@ -0,0 +1,28 @@
+import {makeAutoObservable, observable} from 'mobx';
```
each substore are logical chunks lets name them based on features 

AgentStore , taskStore etc 
 we can have a call to talk about this

mkesavan13 replied:
Hi @arun3528 

I agree. We would need to do that, except that we won't do it right away. At the start, we will keep it limited to just one common store and later when we build the Tasks widget, we will introduce a task store as part of that PR, agent store as part of that PR and so on, although for AgentLogin, we need not have a separate store. Just the common store shall be good to consume. 

Do we want to keep this a non-blocking comment and address it as part of future PRs?

arun3528 replied:
sounds good to me

mkesavan13 commented on package.json:
```diff
@@ -5,10 +5,24 @@
+    "build:providers": "yarn workspace @webex/widget-provider build:src",
+    "build:widgets": "yarn workspace @webex/widget-cc-station-login build:src && yarn workspace @webex/widget-cc-user-state build:src",
+    "samples:build": "yarn workspace @webex/react-samples-app build:src && yarn workspace @webex/web-component-samples-app build:src",
+    "samples:serve-react": "yarn workspace @webex/react-samples-app serve",
+    "samples:serve-wc": "yarn workspace @webex/web-component-samples-app serve"
```
Non-blocking: Can we have one serve command that shows both wc as well as react apps?

Shreyas281299 replied:
https://github.com/user-attachments/assets/1c988e62-719e-4274-b039-c6a21f5dcfd4

This is kind of hacky

What I'm doing?
- Running the react-samples-app in the background (this is at localhost:300)
- Opening the home page 

What happens when we click on the buttons?
- When we click on "Contact Center in a react app", I'm re-directly to localhost:3000
- When we click on "Contact Center widgets as web components" - Im opening the html file that contains the web-component

Problem 
- Not sure if we will be able to run this in github pages
- Not sure if this is a correct way of running both the samples app from the same page

Shreyas281299 replied:
Now that I think about it maybe we should have another application that imports both react and web-component app and then just renders them. And we can run this app using webpack-dev-server

mkesavan13 commented on packages/@webex/widget-cc-user-state/src/user-state/user-state.presentational.tsx:
```diff
@@ -0,0 +1,14 @@
+const UserStatePresentational: React.FunctionComponent<IUserState> = (props) => {
+  return (
+    <>
+      <h1 data-testid="user-state-heading">{props.name}</h1>
```
Why do we have this `data-testid` everywhere? 

Are we also setup for the UI testing?

Shreyas281299 replied:
This I've added for simple unit testing for UI. This is used in the test that checks if the correct prop is rendered. I think that much testing for a basic UI is okay? What do you think?

mkesavan13 commented on packages/@webex/widgets/package.json:
```diff
@@ -33,8 +33,8 @@
     "@webex/components": "1.275.2",
-    "@webex/sdk-component-adapter": "1.112.12",
-    "webex": "2.60.4"
+    "@webex/sdk-component-adapter": "1.112.11",
+    "webex": "2.60.2"
```
This is resetting Parimala's change. Probably a conflict resolution issue. Please check

Shreyas281299 replied:
I missed this. Reverted this change

Shreyas281299 commented on packages/@webex/widget-cc-station-login/src/hooks.ts:
<img width="837" alt="Screenshot 2024-11-11 at 11 02 44 PM" src="https://github.com/user-attachments/assets/bfa9b151-0824-4569-a0ff-ca522814cc36">
The idea behind moving the hooks/helper file outside is that a single widget can have multiple conponents and all these components will use only 1 hooks/helper file. Benefit for this is that we can easily ship the hooks file separately that will provide the business logic and the presentational layer is present inside the folders.
@mkesavan13  please add anything that Ive missed.

Shreyas281299 commented on docs/web-component-samples/src/index.ts:
```diff
@@ -0,0 +1,10 @@
+import {WebStationLogin} from '@webex/widget-cc-station-login';
+import {WebUserState} from '@webex/widget-cc-user-state';
+
+if (!customElements.get('web-user-state')) {
```
Moved this logic inside each widget- now developers only need to import the package and then the custom web components will be available

Shreyas281299 commented on packages/@webex/widget-cc-station-login/tests/station-login/station-login.hooks.ts:
```diff
@@ -0,0 +1,8 @@
+import {useStationLogin} from '../../src/helper';
+
+describe('Station login hooks', () => {
```
Changed

Shreyas281299 commented on packages/@webex/widget-cc-station-login/tsconfig.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "extends": "../../../tsconfig.json",
+    "compilerOptions": {
+      "outDir": "./dist",
+      "declaration": true,
+      "declarationDir": "./dist/types"
+    },
+    "include": [
+      "./src"
+    ]
```
Moved to global config

Shreyas281299 commented on packages/@webex/widget-cc-station-login/webpack.config.js:
```diff
@@ -0,0 +1,12 @@
+const baseConfig = require('../../../webpack.config');
+
+module.exports = merge(baseConfig, {
+  output: {
+    path: path.resolve(__dirname, 'dist'),
+    filename: 'index.js', // Set the output filename to index.js
+    libraryTarget: 'commonjs2',
+  },
```
moved to global config

Shreyas281299 commented on packages/@webex/widgets-store/package.json:
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "@webex/widgets-store",
+  "description": "Webex Contact Center Widgets: Widgets Provider",
```
Updated

Shreyas281299 commented:
> I dont think you have to follow the structure of packages/@webex structure here
> 
> yarn workspace can just relay on the folder structure like widgets/cc-station-login etc
> 
> the npm package where you are going to publish can be @webex/cc-station-login
> 
> The naming looks little messed up i know we did not follow this on the sdk side

Moved the cc-widgets out of @webex folder.

arun3528 commented:
@Shreyas281299 @mkesavan13  i am still not sure about the folder structure and the naming we have to make it little meaningful 

one thing is like naming as its in widgets folder and repo we dont have to add workd widget-cc word all the time could be just cc-stationLogin

Shreyas281299 commented:
> <img alt="Screenshot 2024-11-12 at 10 27 11 AM" width="987" src="https://private-user-images.githubusercontent.com/3895486/385348322-fb0a03fa-d7ae-43ae-8f33-527a9ca4f5f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0ODIxNDQsIm5iZiI6MTczMTQ4MTg0NCwicGF0aCI6Ii8zODk1NDg2LzM4NTM0ODMyMi1mYjBhMDNmYS1kN2FlLTQzYWUtOGYzMy01MjdhOWNhNGY1ZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTNUMDcxMDQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTU5YzAxODAwNGQwMDJkMDFiZDc3NzA4YmQ2OTg1OGI1YzRlMTZlNDgyZmVmNTQzYWM2NGY2MWEzZDExZDUwYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.jvQijnQJCjJ1v72fQ3x_lrmfrhFXyR1KleIK_B0i67I"> @Shreyas281299 @mkesavan13 i am still not sure about the folder structure and the naming we have to make it little meaningful
> 
> one thing is like naming as its in widgets folder and repo we dont have to add workd widget-cc word all the time could be just cc-stationLogin
> 
> you can also change the @webex/widget folder as well if needed

- Moved widgets inside a contact-center folder
- renamed all the widgets to 'cc-X'

mkesavan13 commented on packages/@webex/widget-cc-station-login/src/station-login/index.tsx:
```diff
@@ -0,0 +1,29 @@
+
+if (!customElements.get('widget-cc-station-login')) {
+  customElements.define('widget-cc-station-login', WebStationLogin);
+}
```
We will do this as part of [SPARK-580563](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-580563)

sreenara commented:
:tada: This PR is included in version 1.28.0-alpha.1 :tada:

The release is available on [GitHub release](https://github.com/webex/widgets/releases/tag/1.28.0-alpha.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

sreenara commented:
:tada: This PR is included in version 1.28.0-eft-pipeline.1 :tada:

The release is available on [GitHub release](https://github.com/webex/widgets/releases/tag/1.28.0-eft-pipeline.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

sreenara commented:
:tada: This PR is included in version 1.27.6-eft.1 :tada:

The release is available on [GitHub release](https://github.com/webex/widgets/releases/tag/1.27.6-eft.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

sreenara commented:
:tada: This PR is included in version 1.27.6-eft.1 :tada:

The release is available on [GitHub release](https://github.com/webex/widgets/releases/tag/v1.27.6-eft.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

sreenara commented:
:tada: This PR is included in version 1.28.0-wxcc.1 :tada:

The release is available on [GitHub release](https://github.com/webex/widgets/releases/tag/v1.28.0-wxcc.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

sreenara commented:
:tada: This PR is included in version 1.28.0-ccwidgets.1 :tada:

The release is available on [GitHub release](https://github.com/webex/widgets/releases/tag/v1.28.0-ccwidgets.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

sreenara commented:
:tada: This PR is included in version 1.28.0-ccconnectors.1 :tada:

The release is available on [GitHub release](https://github.com/webex/widgets/releases/tag/v1.28.0-ccconnectors.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
